### PR TITLE
jump links feature TCA-309

### PR DESCRIPTION
**Pull request by lecosson**
https://oat-sa.atlassian.net/browse/ACTP-309

Add invisible jump links to test

**How to test:**
- switch taoAct in TAO instance to `feature/ACTP-305/timer-review-panel-aria` branch
- disable current key navigation plugin to access native key navigation flow
- switch `oat-sa/tao-test-runner-qti-fe` in `taoQtiTest` extension to `feature/ACTP-309/jump-links` branch or use `npm link`
- prepare a test
- login to ACT as a Guest or as user and run this test
- try to acces review panel or other points, or open shortcuts panel, via `jump links` from test using `TAB` key

**related PR's:**
https://github.com/oat-sa/extension-tao-act/pull/1353
https://github.com/oat-sa/extension-tao-testqti/pull/1752

### CODE REVIEW

File: src/provider/qti.js [hunk 1]
```diff
@@ -361,6 +361,85 @@ var qtiProvider = {
...
+                            .on("keydown", closeHandler);
+                    }
+                },
+            };
+            _.forOwn(_jumpLinksBehavior, (linkDescription) => {
+                const link = $(linkDescription.selector);
+                if (link) {
+                    link.on('click', linkDescription.handler);
+                    link.on('keyup', (event) => {
+                        var activationKeys = [32, 13]; // link can be activated by click or enter/space keys
```
Comment (ampaveliev) [inline src/provider/qti.js:435]: const

File: src/provider/qti.js [hunk 2]
```diff
@@ -361,6 +361,85 @@ var qtiProvider = {
...
+                    handler: () => {
+                        const e = findFocusable( areaBroker.getPanelArea() );
+                        e && e.focus();
+                    }
+                },
+                jumpLinkShortcuts: {
+                    selector: '.jump-links-box .jump-link-shortcuts',
+                    handler: () => {
+                        const _shortcutsSelector = ".jump-links-box .shortcuts-list-wrapper";
+                        const _shortcutsList = $(_shortcutsSelector);
```
Comment (ampaveliev) [inline src/provider/qti.js:411]: > Variable names for jQuery elements have to start with $

https://hub.taotesting.com/articles/frontend/architecture/tao-frontend-coding-guide#general-rules

File: src/provider/qti.js [hunk 3]
```diff
@@ -361,6 +361,85 @@ var qtiProvider = {
...
+                    selector: '.jump-links-box .jump-link-teststatus',
+                    handler: () => {
+                        const e = findFocusable( areaBroker.getPanelArea() );
+                        e && e.focus();
+                    }
+                },
+                jumpLinkShortcuts: {
+                    selector: '.jump-links-box .jump-link-shortcuts',
+                    handler: () => {
+                        const _shortcutsSelector = ".jump-links-box .shortcuts-list-wrapper";
```
Comment (ampaveliev) [inline src/provider/qti.js:410]: > Use single quotes for string literals

File: src/provider/layout.tpl [hunk 4]
```diff
@@ -1,5 +1,109 @@
...
+                <li data-control="jump-link-teststatus">
+                    <a role="button" class="jump-link-teststatus" href="#">{{__ "Jump to:"}} <b>{{__ "Test Status and Structure"}}</b></a>
+                </li>
+                <li data-control="jump-link-shortcuts">
+                    <a role="button" class="jump-link-shortcuts" href="#">{{__ "Open Keyboard Shortcuts"}}</a>
+                </li>
+            </ul>
+            <div class="shortcuts-list-wrapper hidden">
+                <div class="shortcuts-list">
+                    <h1 id="keyboard-navigation">Keyboard Navigation</h1>
```
Comment (ampaveliev) [inline src/provider/layout.tpl:23]: Should all strings be translated?

Comment (ampaveliev) [reply src/provider/layout.tpl:23]: Also not sure that it should be h1

File: scss/inc/_action-bars.scss [hunk 5]
```diff
@@ -41,6 +41,71 @@
...
+            &:focus {
+                width: auto;
+                height: auto;
+                padding: 30px !important;
+                background: $uiClickableDefaultBg;
+                color: $textColor !important;
+                border: 3px solid $textColor;
+                outline: none;
+            }
+            &:focus:hover {
```
Comment (ampaveliev) [inline scss/inc/_action-bars.scss:76]: No hover state?

File: scss/inc/_action-bars.scss [hunk 6]
```diff
@@ -41,6 +41,71 @@
         }
     }
 
+    .jump-links-box {
+        display: block;
+        width: 0px;
```
Comment (ampaveliev) [inline scss/inc/_action-bars.scss:46]: You can use `width: 0`.  This is considered best practice

File: scss/inc/_action-bars.scss [hunk 7]
```diff
@@ -41,6 +41,71 @@
...
+
+        [class^=jump-link] {
+            z-index: 1002;
+            position: fixed;
+            display: block;
+            overflow: hidden;
+            top: 10px;
+            left: 10px;
+            width: 0px;
+            padding: 0px !important;
```
Comment (ampaveliev) [inline scss/inc/_action-bars.scss:64]: Can it be styled without `!important`?

Comment (lecosson) [reply scss/inc/_action-bars.scss]: outdated

File: scss/inc/_action-bars.scss (hunk 5 above)
Comment (lecosson) [reply scss/inc/_action-bars.scss]: outdated

File: src/provider/layout.tpl (hunk 4 above)
Comment (lecosson) [reply src/provider/layout.tpl]: outdated

File: src/provider/qti.js (hunk 3 above)
Comment (lecosson) [reply src/provider/qti.js]: outdated

File: src/provider/qti.js (hunk 2 above)
Comment (lecosson) [reply src/provider/qti.js]: outdated

File: src/provider/qti.js (hunk 1 above)
Comment (lecosson) [reply src/provider/qti.js]: outdated

File: src/plugins/content/accessibility/jumplinks/jumplinks.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+ */
+export default function jumplinksFactory(config) {
+
+    /**
+     * @typedef {Object} jumplinksBox
+     */
+    const jumplinksBox = component({}, defaults )
+    .on('render', function() {
+        // handle related Jump Links
+        const _jumpLinksBehavior = {
```
Comment (AntonTsymuk) [inline src/plugins/content/accessibility/jumplinks/jumplinks.js:51]: maybe it is better to us array here?
not sure if jumpLinksBehavior suitable name here. i do not see any think related to behavior

File: src/plugins/content/accessibility/jumplinks/plugin.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        self.jumplinks = jumplinksFactory({})
+            .on('render', function() {
+                handleJumpLinks();
+            });
+        self.shortcuts = shortcutsFactory({});
+
+        function handleJumpLinks() {
+            const _jumpLinksBehavior = {
+                jumpLinkQuestion: {
+                    selector: '.top-action-bar .jump-links-box [data-jump=question] ',
```
Comment (AntonTsymuk) [inline src/plugins/content/accessibility/jumplinks/plugin.js:66]: do you need this selectors?

File: src/plugins/content/accessibility/jumplinks/plugin.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+                            .off("click", closeHandler)
+                            .on("click", closeHandler);
+                        $(window)
+                            .off("keydown", closeHandler)
+                            .on("keydown", closeHandler);
+                    }
+                },
+            };
+            _.forOwn(_jumpLinksBehavior, (linkDescription) => {
+                self.jumplinks.on(linkDescription.eventName, () => {
```
Comment (AntonTsymuk) [inline src/plugins/content/accessibility/jumplinks/plugin.js:119]: you do not need so many events. two will be enough:
- one for jump links
- one for shortcuts

please pass additional arguments to the shortcuts event and use area broker based on it
```
const mapJumpToAreaBroker = {
teststatus: 'getPanelArea',
...
}

on('...', (jump) => {
const $element = self.getAreaBroker()[mapJumpToAreaBroker[jump]]();
...
}
```

File: src/plugins/content/accessibility/jumplinks/plugin.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                    handler: () => {
+                        const e = findFocusable(self.getAreaBroker().getPanelArea() );
+                        e && e.focus();
+                    }
+                },
+                jumpLinkShortcuts: {
+                    selector: '.top-action-bar [data-jump=shortcuts]',
+                    eventName: 'shortcuts',
+                    handler: () => {
+                        self.shortcuts.show();
```
Comment (ampaveliev) [inline src/plugins/content/accessibility/jumplinks/plugin.js:101]: No needs to use `self` in the arrow functions.

File: src/plugins/content/accessibility/jumplinks/plugin.js [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+                });
+            });
+        }
+    },
+
+    /**
+     * Called during the runner's render phase
+     */
+    render: function render() {
+        const self = this;
```
Comment (ampaveliev) [inline src/plugins/content/accessibility/jumplinks/plugin.js:131]: Self variable is redundant

File: src/plugins/content/accessibility/jumplinks/jumplinks.js [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            },
+            jumpLinkToolbox: {
+                selector: '.top-action-bar .jump-links-box .jump-link-toolbox',
+                handler: () => {
+                    const e = findFocusable( areaBroker.getToolboxArea() );
+                    e && e.focus();
+                }
+            },
+            jumpLinkTeststatus: {
+                selector: '.top-action-bar .jump-links-box .jump-link-teststatus',
```
[the remark this answers is not in the document]
Comment (lecosson) [reply src/plugins/content/accessibility/jumplinks/jumplinks.js]: refactored

File: src/plugins/content/accessibility/jumplinks/jumplinks.js (hunk 8 above)
Comment (lecosson) [reply src/plugins/content/accessibility/jumplinks/jumplinks.js:51]: refactored

File: src/plugins/content/accessibility/jumplinks/plugin.js (hunk 9 above)
Comment (lecosson) [reply src/plugins/content/accessibility/jumplinks/plugin.js:66]: removed on refactoring

File: src/plugins/content/accessibility/jumplinks/plugin.js (hunk 10 above)
Comment (lecosson) [reply src/plugins/content/accessibility/jumplinks/plugin.js:119]: 👌

File: src/plugins/content/accessibility/jumplinks/plugin.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+                if ($elementGetter) {
+                    const $focusable = findFocusable($elementGetter());
+                    $focusable && $focusable.focus();
+                }
+            });
+
+            self.jumplinks.on('shortcuts', () => {
+                self.jumplinks.getElement().find(':focus').blur();
+                self.shortcuts.show();
+                const closeHandler = function(event) {
```
[the remark this answers is not in the document]
Comment (lecosson) [reply src/plugins/content/accessibility/jumplinks/plugin.js:78]: moved

File: src/plugins/content/accessibility/jumplinks/jumplinks.js [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+const defaults = {
+};
+
+/**
+ * Creates and initialize the jumplinks component.
+ * Please not the component IS NOT rendered.
+ * You'll have to render it by yourself.
+ *
+ * @param {Object} config
+ * @param {Object[]} [config.areaBroker] - test page area broker
```
Comment (AntonTsymuk) [inline src/plugins/content/accessibility/jumplinks/jumplinks.js:40]: redundad param

File: src/plugins/content/accessibility/jumplinks/plugin.js [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+ * @author aliaksandr paliakou <lecosson@gmail.com>
+ */
+
+import $ from 'jquery';
+import pluginFactory from 'taoTests/runner/plugin';
+import jumplinksFactory from "./jumplinks";
+import shortcutsFactory from "./shortcuts";
+
+/**
+ * adds rendered jump links support
```
Comment (AntonTsymuk) [inline src/plugins/content/accessibility/jumplinks/plugin.js:29]: what is this comment about?

Comment (lecosson) [reply src/plugins/content/accessibility/jumplinks/plugin.js:29]: deprecated lines. removed

File: src/plugins/content/accessibility/jumplinks/plugin.js [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+ * @author aliaksandr paliakou <lecosson@gmail.com>
+ */
+
+import $ from 'jquery';
+import pluginFactory from 'taoTests/runner/plugin';
+import jumplinksFactory from "./jumplinks";
+import shortcutsFactory from "./shortcuts";
+
+function findFocusable(targetElement) {
+    const $elem = $(targetElement)
```
Comment (ampaveliev) [inline src/plugins/content/accessibility/jumplinks/plugin.js:29]:
```js
return $(targetElement)
  .find('input, select, a[href], textarea, button, [tabindex]')
  .toArray()
  .filter((el) => (el.tabIndex >= 0 && !el.disabled && el.offsetParent ) )
  .find((el) => (typeof el.focus === 'function') );
```

File: src/plugins/content/accessibility/jumplinks/plugin.js [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+            question: 'getContentArea',
+            navigation: 'getNavigationArea',
+            toolbox: 'getToolboxArea',
+            teststatus: 'getPanelArea',
+        };
+        self.jumplinks = jumplinksFactory({})
+            .on('render', handleJumpLinks);
+        self.shortcuts = shortcutsFactory({});
+
+        function handleJumpLinks() {
```
Comment (ampaveliev) [inline src/plugins/content/accessibility/jumplinks/plugin.js:68]: Prefer arrow functions and remove `self` variable

Comment (ampaveliev) [reply src/plugins/content/accessibility/jumplinks/plugin.js:68]: Also hoisted function might not be obvious

File: src/plugins/content/accessibility/jumplinks/shortcuts.tpl [hunk 19]
```diff
@@ -0,0 +1,85 @@
+<div class="shortcuts-list-wrapper hidden">
+    <div class="shortcuts-list">
+        <h1 id="keyboard-navigation">Keyboard Navigation</h1>
```
Comment (ampaveliev) [inline src/plugins/content/accessibility/jumplinks/shortcuts.tpl:3]: Should the string in the template be translated? Also using h1 should be discussed with UX

File: src/plugins/content/accessibility/jumplinks/plugin.js (hunk 17 above)
Comment (lecosson) [reply src/plugins/content/accessibility/jumplinks/plugin.js:29]: refactored to fix support keynavigation plugin

File: src/plugins/content/accessibility/jumplinks/plugin.js (hunk 18 above)
Comment (lecosson) [reply src/plugins/content/accessibility/jumplinks/plugin.js]: refactored

**Pull request merged**